rfd/0034-clients-rotate-certs.md

```diff
@@ -0,0 +1,187 @@
+---
+authors: Brian Joerger (bjoerger@goteleport.com)
+state: draft
+---
+
+# RFD 34 - Clients Certificate Rotation
+
+## What
+
+Provide a way for long living automated clients to retrieve new User certificates as needed.
+
+## Why
+
+Presently, there is no good way to keep client certificates secure and valid for long durations. In order to gain access for an extended period of time, one must retrieve a set of long lived certificates. There are a few security and UX issues with this approach.
+
+- Long lived certificates are less secure than short lived certificates.
+- The User CA may be rotated due to a security breach or a scheduled rotation. When this happens, all User certificates will lose access, after a grace period, and need to be manually refreshed.
+- Eventually, long lived certificates will expire after their TTL. Then they need to be manually refreshed, which puts a strain on administrators and could lead to downtime if neglected.
+
+## Details
+
+Cert agent will be a new standalone client with the responsibility of continuously renewing a set of certificates before they lose access. Before explaining how the agent will function, we need to create a new abstraction for certificates to enable the agent.
+
+### Cert Store
+
+The agent will interact with certificates through its Cert Store.
+
+A Cert Store can be used to:
+ - authenticate a client
+ - store renewed certificates
+ - alert when the store's certificates are altered
+ - provide its certificates expiration time
+
+```go
+type Store interface {
```

Comment:
nit: Store is super generic, unless it's in credentials package, I'd use CredStore

Reply:
Agreed, I'm planning on putting in a `cert` package, so it'll be `cert.Store` and `cert.Agent`. I think that certs/certificates is preferable to credentials. It might be worth it to rename `client.Credentials` to `client.Certs` as well, which can be done without breaking backwards compatibility.